scripts/components.py

```diff
@@ -0,0 +1,153 @@
+    filename = "{name}.yml".format(name=name)
+    filepath = os.path.join(directory, filename)
+
+    return load_data(filepath, component_schema)
+
+
+def save_component(component, directory):
+    filename = "{name}.yml".format(name=component["name"])
+    filepath = os.path.join(directory, filename)
+
+    save_data(filepath, component, component_schema)
+
+
+def update_component(existing, name, repo_url, is_product):
+    if existing:
+        component = deepcopy(existing)
+    else:
+        component = {
+            "name": None,
+            "repo_url": None,
+            "is_product": None,
+            "releases": [],
+        }
+    new = {
+        k: v
+        for k, v in (
+            ("name", name),
+            ("repo_url", repo_url),
+            ("is_product", is_product)
+        )
+        if v is not None
+    }
+
+    component.update(**new)
+
+    return component_schema.validate(component)
+
+
+def update_releases(component, series_name, version, sha):
+    component = deepcopy(component)
+    release = {
+        "version": version,
+        "sha": sha,
+    }
+
+    for r in component["releases"]:
+        if r["series"] == series_name:
+            series_versions = r["versions"]
+            break
+    else:
+        series_versions = []
+        component["releases"].append(
+            {
+                "series": series_name,
+                "versions": series_versions,
+            }
+        )
+
+    series_versions.append(release)
+
+    return component_schema.validate(component)
+
+
+def parse_args(args):
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument("--component-name", required=True)
+    parser.add_argument("--component-dir", default="./")
+
+    subparsers = parser.add_subparsers(dest="subparser")
+    subparsers.required = True
+
+    c_parser = subparsers.add_parser("component")
+
+    c_parser.add_argument("--repo-url", type=repo_url_schema.validate)
+    c_parser.add_argument("--is-product", action="store_true", default=False)
+
+    r_parser = subparsers.add_parser("release")
+
+    r_parser.add_argument(
+        "--version", type=version_id_schema.validate, required=True
+    )
+    r_parser.add_argument(
+        "--sha", type=version_sha_schema.validate, required=True
+    )
+    r_parser.add_argument("--series-name", required=True)
+
+    return vars(parser.parse_args(args))
+
+
+def main(raw_args=None):
+    try:
+        kwargs = parse_args(raw_args)
+
+        subparser = kwargs.pop("subparser")
+        component_name = kwargs.pop("component_name")
+        component_dir = kwargs.pop("component_dir")
+
+        existing_component = load_component(component_name, component_dir)
+
+        if subparser == "component":
+            updated_component = update_component(
+                existing_component, name=component_name, **kwargs
+            )
+        elif subparser == "release":
+            updated_component = update_releases(existing_component, **kwargs)
```

Comment:
Should you ensure the component exists before you try to update it?

Reply:
Not necessarily but that didn't fail with a nice message so I've added something for now although I may want to rework that area.
```
$ component --releases-dir . release --component-name foobar --version 1.1.1 --series-name blah --sha 0000000000000000000000000000000000000000
Component 'foobar' does not exist.

```